File: strykerjs-node-test-poc/src/validator.ts
export type ValidationResult = { valid: true } | { valid: false; reasons: string[] };

export function checkLength(password: string): string | undefined {
  return password.length < 8 ? 'Password must be at least 8 characters' : undefined;
}

export function checkDigit(password: string): string | undefined {
  return /\d/.test(password) ? undefined : 'Password must contain a digit';
}

export function checkUpper(password: string): string | undefined {
  return /[A-Z]/.test(password) ? undefined : 'Password must contain an uppercase letter';
}

export function validatePassword(password: string): ValidationResult {
  const reasons = [checkLength(password), checkDigit(password), checkUpper(password)].filter(
    (reason): reason is string => reason !== undefined,
  );
  return reasons.length === 0 ? { valid: true } : { valid: false, reasons };
}

export function classifyAge(age: number): string {
  if (age < 0) return 'invalid';
  if (age < 13) return 'child';
  if (age < 20) return 'teen';
  if (age < 65) return 'adult';
  return 'senior';
}

File: stryker4s-poc/src/main/scala/poc/Validator.scala
package poc

sealed trait ValidationResult
case object Valid extends ValidationResult
final case class Invalid(reasons: List[String]) extends ValidationResult

object Validator {

  def checkLength(password: String): Option[String] =
    if (password.length < 8) Some("Password must be at least 8 characters") else None

  def checkDigit(password: String): Option[String] =
    if (!password.exists(_.isDigit)) Some("Password must contain a digit") else None

  def checkUpper(password: String): Option[String] =
    if (!password.exists(_.isUpper)) Some("Password must contain an uppercase letter") else None

  def validatePassword(password: String): ValidationResult = {
    val reasons = List(checkLength(password), checkDigit(password), checkUpper(password)).flatten
    if (reasons.isEmpty) Valid else Invalid(reasons)
  }

  def classifyAge(age: Int): String = age match {
    case a if a < 0  => "invalid"
    case a if a < 13 => "child"
    case a if a < 20 => "teen"
    case a if a < 65 => "adult"
    case _           => "senior"
  }
}
